### PR TITLE
Add reset to QASM parser

### PR DESCRIPTION
**Pull request by rscircus**
This PR solves #6704 

Note / 20240914:

After further inspection, I wonder why resets are implemented quite different to regular gates, despite following a somewhat similar terminology (`cirq.R` or `cirq.reset`) however, returning a [ResetChannel](https://quantumai.google/reference/python/cirq/ResetChannel). It appears to me that the reset is more similar to a measurement than compared to say a Hadamard gate.

### CODE REVIEW

**Comment by rscircus** [comment]
Rebased on top of main.

**Comment by rscircus** [comment]
Squashing everything into one commit after we finish review. Thus I leave the generic 'Fix comments & typos' as it is for now.

**Comment by rscircus** [comment]
Hey guys, you have any feedback on the solution proposal here?

File: cirq-core/cirq/contrib/qasm_import/_parser.py [hunk 1]
```diff
@@ -499,6 +501,15 @@ def p_measurement(self, p):
...
         ]
 
+    # reset operations
+    # reset : RESET qarg
+
+    def p_reset(self, p):
+        """reset : RESET qreg ';'"""
+        qreg = p[2]
+
+        p[0] = [ops.ResetChannel().on(qreg[i]) for i in range(len(qreg))]
```
Comment (natestemen) [inline cirq-core/cirq/contrib/qasm_import/_parser.py:511]: Should there be some sort of check here to ensure the reset is applied on a qubit that is actually defined in the quantum register? E.g. should some error be raised when trying to parse the following QASM program?
```qasm
OPENQASM 2.0;
include "qelib1.inc";
qreg q[1];
reset q[2];
```

**Review by natestemen — APPROVED** [review]
My review doesn't mean much, since I'm not very familiar with this codebase, but the diff looks reasonable! Hope this can get reviewed by a cirq maintainer soon :)

**Comment by dstrain115** [comment]
It looks like there is an error in the parser:

`ERROR: /home/runner/work/Cirq/Cirq/cirq-core/cirq/contrib/qasm_import/_parser.py:508: Symbol 'qreg' used, but not defined as a token or a rule`

Also, please run check/formatting in order to pass the format check.

**Comment by gluonhiggs** [comment]
> It looks like there is an error in the parser:
> 
> `ERROR: /home/runner/work/Cirq/Cirq/cirq-core/cirq/contrib/qasm_import/_parser.py:508: Symbol 'qreg' used, but not defined as a token or a rule`
> 
> Also, please run check/formatting in order to pass the format check.

Hello everyone, is there anything remaining to address in this PR before it can be finalized?

**Comment by pavoljuhas** [comment]
@rscircus - thank you for contributing this and apologies for slow response.

I have fixed the typo in parser definition and complaints from format checker.
I also added a test of the `reset` instruction with OpenQASM 3.

Please take a look if you are fine with the PR as is.  LGTM from my POV.

**Review by pavoljuhas — APPROVED** [review]

File: cirq-core/cirq/contrib/qasm_import/_parser.py (hunk 1 above)
Comment (pavoljuhas) [reply cirq-core/cirq/contrib/qasm_import/_parser.py:538]: The example program already raises a QasmException at
https://github.com/quantumlib/Cirq/blob/e1e1730e8ab41cd6bcfc655b1a78993442797d7b/cirq-core/cirq/contrib/qasm_import/_parser.py#L459-L464

**Comment by rscircus** [comment]
Thanks. I'll pull at the weekend and give it a final run.

**Comment by pavoljuhas** [comment]
> Thanks. I'll pull at the weekend and give it a final run.

I have merged, hope it is OK.  LMK if you see anything amiss and we can address it in a follow-up PR.

**Comment by rscircus** [comment]
Thanks, @pavoljuhas

**Pull request merged**